contracts/SurplusBeneficiary.sol

```diff
@@ -0,0 +1,122 @@
+import { SurplusBeneficiaryStorageV1 } from "./storage/SurplusBeneficiaryStorage.sol";
+import { ISurplusBeneficiary } from "./interface/ISurplusBeneficiary.sol";
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is ISurplusBeneficiary, ReentrancyGuard, Ownable, SurplusBeneficiaryStorageV1 {
+    using Address for address;
+    using SafeMath for uint256;
+
+    constructor(
+        address tokenArg,
+        address feeDistributorArg,
+        address treasuryArg,
+        uint24 treasuryPercentageArg
+    ) ReentrancyGuard() Ownable() {
+        // SB_TANC: token is not a contract
+        require(tokenArg.isContract(), "SB_TANC");
+        _token = tokenArg;
+
+        setFeeDistributor(feeDistributorArg);
+        setTreasury(treasuryArg);
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20(_token).approve(feeDistributorArg, type(uint256).max);
+    }
+
+    //
+    // PUBLIC ONLY-ADMIN NON-VIEW
+    //
+
+    function setFeeDistributor(address feeDistributorArg) public onlyOwner {
+        // SB_FDNC: feeDistributor address is not contract
+        require(feeDistributorArg.isContract(), "SB_FDNC");
+
+        // SB_TNM: token is not match
+        require(IFeeDistributor(feeDistributorArg).token() == _token, "SB_TNM");
+
+        address oldFeeDistributor = _feeDistributor;
+        _feeDistributor = feeDistributorArg;
+        emit FeeDistributorChanged(oldFeeDistributor, feeDistributorArg);
+    }
+
+    function setTreasury(address treasuryArg) public onlyOwner {
+        // SB_TZ: treasury address is zero
+        require(treasuryArg != address(0), "SB_TZ");
+
+        address oldTreasury = _treasury;
+        _treasury = treasuryArg;
+        emit TreasuryChanged(oldTreasury, treasuryArg);
+    }
+
+    function setTreasuryPercentage(uint24 treasuryPercentageArg) public onlyOwner {
+        // SB_TPO: treasury percentage out of bound
+        require(treasuryPercentageArg <= 1e6, "SB_TPO");
+
+        uint24 oldTreasuryPercentage = _treasuryPercentage;
+        _treasuryPercentage = treasuryPercentageArg;
+        emit TreasuryPercentageChanged(oldTreasuryPercentage, treasuryPercentageArg);
+    }
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    /// @inheritdoc ISurplusBeneficiary
+    function dispatch() external override nonReentrant {
+        address token = _token;
+
+        uint256 tokenAmount = IERC20(token).balanceOf(address(this));
+
+        // SB_TAZ: token amount is zero
+        require(tokenAmount != 0, "SB_TAZ");
+
+        // note the ratio is in `1e6` format
+        uint256 tokenAmountToTreasury = FullMath.mulDiv(tokenAmount, _treasuryPercentage, 1e6);
+
+        // transfer to treasury first, because FeeDistributor.burn() will transfer all balance from SurplusBeneficiary
+        SafeERC20.safeTransfer(IERC20(token), _treasury, tokenAmountToTreasury);
+        IFeeDistributor(_feeDistributor).burn(token);
```

Comment:
although it will never return false, but since burn returns boolean, do you think we should add `require(fd.burn())`?